---
title: ""
author: "Hana Parizkova & Deepak Tanwar"
date: "<b>Created on:</b> 2018-12-21 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  html_document:
    theme: spacelab
    highlight: pyg
    keep_md: no
    number_sections: no
    fig_width: 8.5
    fig_height: 11
    fig_caption: true
    df_print: paged
    code_folding: show
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
---

```{r, warning=F, message=FALSE, echo=FALSE}
```


# Introduction


# Methods
## Data
Two data sets were used.

One, representing negative control (with no expected DMRs), contained six methylation profiling samples from normal human dendritic cells ([data in GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE64177)). The samples were artificially divided into two groups of three samples, group 1 consisting of samples [GSM1565940](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1565940), [GSM1565944](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1565944), and [GSM1565948](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1565948), group 2 consisting of samples [GSM1565942](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1565942), [GSM1565946](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1565946), and [GSM1565950](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1565950). To obtain reasonable running times, only methylation data from chromosome 18 were used.

The second set of data was created from the negative control by adding 100 simulated DMRs using `simDMRS` function from R package `dmrseq`. 

[Data download and filtering](https://github.com/sta426hs2018/sta426-project-dmr-comparison/blob/master/data/data_download_and_filter.html)

## The methods used for comparison
In total 7 different methods were used for the comparison. Their main features are summarized in the following table:

| Method     | Preprocessing + statistical test used                                                   | p-value estimate given | CpG / non CpG | Implementation language | DOI                           | Availability                                                                                         |
|------------|-----------------------------------------------------------------------------------------|------------------------|---------------|-------------------------|-------------------------------|------------------------------------------------------------------------------------------------------|
| bsseq      | local-likelihood smoothing, t-test similar statistics                                   | no                     | CpG only      | R                       | 10.1186/gb-2012-13-10-r83     | [Bioconductor package bsseq](https://bioconductor.org/packages/release/bioc/html/bsseq.html)         |
| CGmapTools | unpaired t-test                                                                         | yes                    | CpG only      | C, C++, Python          | 10.1093/bioinformatics/btx595 | [binary package](https://cgmaptools.github.io/)                                                      |
| defiant    | Weighted Welch Expansion                                                                | yes                    | CpG only      | Shell                   | 10.1186/s12859-018-2037-1     | [binary package](https://github.com/hhg7/defiant)                                                    |
| DMRcaller  | no preprocessing, or pooling in bins, or kernel smoothing; Fisher's exact or Score test | yes                    | all contexts  | R                       | 10.1093/nar/gky602            | [Bioconductor package DMRcaller](https://bioconductor.org/packages/release/bioc/html/DMRcaller.html) |
| dmrseq     | smoothing; generalized least squares                                                    | yes                    | CpG only      | R                       | 10.1093/biostatistics/kxy007  | [Bioconductor package dmrseq](https://bioconductor.org/packages/release/bioc/html/dmrseq.html)       |
| methPipe   | Fisher's exact test                                                                     | no                     | CpG only      | C, C++                  | 10.1371/journal.pone.0081148  | [binary package](http://smithlabresearch.org/software/methpipe/)                                     |
| methylKit  | Fisher's exact test, or logistic regression with tiling                                 | yes                    | all contexts  | R                       | 10.1186/gb-2012-13-10-r87     | [Bioconductor package methylKit](https://bioconductor.org/packages/release/bioc/html/methylKit.html) |

### `bsseq` analysis
link to bsseq

### `CGmapTools` analysis
link to CGmapTools

### `defiant` analysis
link to defiant

### `DMRcaller` analysis
link to DMRcaller

### `dmrseq` analysis
link to dmrseq

### `methPipe` analysis
link to methPipe

### `methylKit` analysis
link to methylKit


# Results
The reported DMRs were (if possible) filtered for only those that contain at least 10 CpGs and the mean difference between the two groups is at least 0.1 (the simulated regions had effect sizes between 0.163 and 0.450).

# Discussion and Conclusion


# SessionInfo

```{r}
devtools::session_info()
```
